### PR TITLE
serve readthedocs

### PR DESCRIPTION
**Pull request by dragazo**
I don't know if this is the direction you wanted to go with hosting service documentation, but I took a swing at it. Related to #3059.

This PR currently just looks through each service on startup and checks if it has a `docs/` directory. If so, it goes into the directory and runs the shell command `make html`. The motivation for this being that this is the default scheme for readthedocs, and even if you use another tool you just need to add a makefile that invokes whatever you used. It does this asynchronously, and on success binds it to `/docs/services/<service-name>`. I tested it with PhoneIoT (`/docs/services/phone-iot`), which is currently the only service with this doc setup.

I haven't added this yet, but I could expose a little API so you can check if a service has docs (and if so, the url), and we could modify `browser` to check for that and have a `docs...` context option, similar to `help...`.

Note that if this is used, the server would need the readthedocs compile chain installed (not too large I don't think). Also, we should pick a consistent style. For PhoneIoT I used `sphinx_rtd_theme` because it's made by the main team and it looks very modern.

### CODE REVIEW

**Comment by brollb** [comment]
Nice! Thanks for taking the initiative :) This has been on my wishlist for a while and will be great to have!

**As a little background on what I had been thinking (subject to change, of course!):**

When moving to use readthedocs for consolidating all the documentation, I had been imagining that the documentation for services would just be one section. Other sections would include many [guides from netsblox.org](https://netsblox.org/help) and [developer documentation](https://github.com/NetsBlox/NetsBlox/wiki).

Since we already know a bit about the services from the documentation, it would be great if a lot of the service documentation was automatically generated from the help messages and method signatures. Ideally, services could add supplementary documentation (I have been imagining as additional pages/subsections) for custom information. Documentation about the mobile app or service-specific walkthroughs would be a couple example use cases.

I had also been imagining that we would just host the documentation on readthedocs. However, I prefer the approach taken here as it can ensure separate docs for editor and dev (which may have different services available). 

**Next Steps:**
The easiest path forward might actually be to:
- Move the main chunk of the documentation to `src/server/docs` or `src/server/readthedocs` or something.
- Generate the services documentation as part of the `npm run build` or as a `prestart` command. This should probably generate rst files from the service documentation (as well as copying over custom files) and place them in the main documentation directory.
- Build the documentation (probably as part of the previous step/npm command).
- Serve the docs like you currently are from the main server. There is probably a better url like `/docs/` rather than `/docs/services`.

If you aren't feeling as ambitious, we could probably get by with just a consolidated documentation for the server and then manually create all the docs for now. It wouldn't be quite as nice but it would work.

A couple additional minor comments:
- The services should not be able to use their own themes
- When possible, errors when building the documentation should be hard errors that kill the server. Documentation is only updated during development and it would be nice to correct errors before merging changes to master.

**Comment by dragazo** [comment]
Ok, I think the latest commit gets us pretty close. I consolidated the docs into `src/server/docs` (bound to `/docs`), and have set it up so that we can add all the services there and they'll all be part of the same document. So you can cross-reference between services, if you want. Also, because they're all the same document, a single theme is applied universally to all of them.

I also changed the docs for PhoneIoT to use the jsdoc directly, so it's super short in the actual `rst` files. This is through the `jsdoc` npm package and `sphinx-js` python package. Each function is specified manually like `.. autofunction:: etc.`. I couldn't find a way to automatically grab all the exposed RPCs to auto-populate them, but given that there could be multiple pages for a given service it's not clear where to put the extras, anyway.

But at least adding docs for a new service is easy. Just add a symlink to the `js` file in `src/server/docs`, make a new directory for the service in `src/server/docs/content`, and add your service-specific rst files. Then add a reference to the service to the `src/server/docs/content/index.rst` file and it'll automatically link everything.

The docs are still build at server start, but I changed it so that compile failure kills the server by unhandled exception.

**Comment by brollb** [comment]
Looking better! It looks like the developer needs to still write the documentation file (though they can use the generated JSDoc). It would be nice if the documentation that you have added for PhoneIoT could be completely generated from the service (with the exception of the "Handling Errors" section which should be shared among the services anyway). To make this happen, I think we need a couple things:
- RPCs can be organized by category automatically
- Categories should be able to have descriptions. These can be placed at the top of the page.

We may need to generate the docs from the service metadata rather than using the JSDoc info. The metadata is available as the `metadata` attribute on the [services worker](https://github.com/NetsBlox/NetsBlox/blob/master/src/server/services/services-worker.js).

Ideally, this will be able to automatically generate documentation for all the existing services. We will probably want to add some tests in the future which test that the service has documentation for each RPC, etc...

**Comment by dragazo** [comment]
Yeah, tbh the autogeneration from jsdoc was annoying my anyway because it was formatting types with multiple generic parameters wrong, and it also made duck typed fields hard to read. In the latest commit I just wrote some code to compile the metadata directly into rst in a temporary directory, then build it into html.

I did this in a way that still allows a good bit of customization. I made some replacement tokens: `>>>RPCS<<<` in an rst file will expand to all of the rpcs, `>>>CATS<<<` gives all the categories, etc. If you don't provide an rst file for a service, it makes one for you. Also, it splits up RPCs by category for organization. So if you do `npm run postinstall` you'll have docs for every service.

**Review by brollb — CHANGES_REQUESTED** [review]
Nice work. It is shaping up nicely! A couple comments:
- I am getting an error when running the build script. It appears that everything is working but is throwing an error nonetheless.
- It is great to see that docs are generated for every service but it might be nice to group them by category like the call RPC block does. (Note that one service may belong to multiple categories.)
- It might be nice to refactor it slightly so the document generation can be tested with unit tests. One approach would be to view the task as 2 steps: JSON transformation then document generation via a (lodash) template. The first function would accept a JSON (`ServicesAPI.services.metadata` by default) and convert it to a format that can be easily dropped into a template. This would also lend itself nicely to when we update it to include only the services defined statically on the filesystem since it would just be providing a different metadata JSON. Anyway, this isn't required but might be nice to make the code cleaner and more conducive to automated tests.

Here is the error I am currently seeing on postinstall:
```
generating indices... genindex done
writing additional pages... search done
copying static files... done
copying extra files... done
dumping search index in English (code: en)... done
dumping object inventory... done
build succeeded, 1 warning.

The HTML pages are in _build/html.

    at /home/irishninja/projektek/netsblox/bin/build.js:221:24
    at ChildProcess.exithandler (child_process.js:286:7)
    at ChildProcess.emit (events.js:223:5)
    at ChildProcess.EventEmitter.emit (domain.js:475:20)
    at maybeClose (internal/child_process.js:1021:16)
    at Socket.<anonymous> (internal/child_process.js:430:11)
    at Socket.emit (events.js:223:5)
    at Socket.EventEmitter.emit (domain.js:475:20)
    at Pipe.<anonymous> (net.js:664:12)
npm ERR! code ELIFECYCLE
npm ERR! errno 197
npm ERR! netsblox@1.34.0 postinstall: `node bin/build.js`
npm ERR! Exit status 197
npm ERR! 
npm ERR! Failed at the netsblox@1.34.0 postinstall script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     /home/irishninja/.npm/_logs/2021-04-23T11_43_01_262Z-debug.log
```

**Comment by dragazo** [comment]
Last two commits should address those first two points, but I haven't started on the third yet. It's at least in a pretty usable form now, but you're right, the testability would be nice. I think the way I have it now I'm basically generating what the separate json repr would be, so I just need to split it away from the file I/O.

**Comment by brollb** [comment]
Nice. Three more comments:
- I am wondering if it would be nicer to have any custom docs be in the same directory as the service code (maybe nested under procedures/<service>/docs). My concern with the current approach is that when a new developer/intern is working on a service, they may need to make changes in two places. I am also worried the docs could be easily overlooked when changes are made to am existing (potentially dev) service. Not necessary to address for this PR but wanted to mention it.
- It would be nice if the services could be nested by category in the left pane, too.
- I just wanted to confirm that it should be easy to add non-service related docs too, right? (I would dig through the code but am writing this from my phone so a little more awkward!)

Looking forward to getting this in! Then we can migrate the existing guides from netsblox.org here and the GitHub wiki!

**Comment by dragazo** [comment]
Having custom docs in the service directory was how it worked with the previous version using the (ironically) manual `autofunction`. In the latest version using metadata directly, I moved it to the docs directory because there's still that issue of not knowing the fs name given the service name. But I think I can add the fs path to the metadata object in the jsdoc extractor, and that should do it. That would also let me do the fs service only filtering. I'll give that a try tonight.

**Comment by brollb** [comment]
Cool. Another option would be to add an optional flag to the initialize method that allowed you to only load fs services. (I prefer this approach but could be persuaded :) )

Get Outlook for Android<https://aka.ms/AAb9ysg>

________________________________

**Comment by brollb** [comment]
Wait, that doesn't really solve the main problem :) That approach sounds good!

Get Outlook for Android<https://aka.ms/AAb9ysg>

________________________________

**Comment by dragazo** [comment]
Cool, looks like fs vs db services had a separate loader, and fs took the module path, so I added a field to keep the info around for later. If a service is an fs service, it looks for a `docs` folder in its directory. I set up the fs-only filtering as well. There are a set of filter functions for services (currently `all` (default) or `fsonly`). You can specify the filter with the `DOCS_SERVICE_FILTER` environment variable.

**Comment by brollb** [comment]
Nice work. Two last things and then I think it should be good to go:
- Could you make sure the postinstall script doesn't fail on github actions? I am not sure why it doesn't seem to be running now but I expect building the docs to fail since I don't believe python is installed currently.
- Could you group the services in the left pane by category?

It looks like there is also an opportunity for some minor code cleanup - particularly around generating the documentation for services. Generally, I find templates much easier to read like [here](https://github.com/NetsBlox/NetsBlox/blob/bcae54822bb6c5ae61965143f39372696e196ead/src/server/api/rest/oauth/index.html.ejs) and provides a little bit of a separation of concerns. This isn't blocking the merge by any means but I thought it was worth mentioning.

**Comment by dragazo** [comment]
>Could you make sure the postinstall script doesn't fail on github actions? I am not sure why it doesn't seem to be running now but I expect building the docs to fail since I don't believe python is installed currently.

Would it suffice to just add python to `docker-compose`? (I haven't done much with github actions, but I think it uses docker.)

>Could you group the services in the left pane by category?

For me, they already are, but you have to click them to expand. Is that not what you're seeing? Or did you mean you want them to show up with the services under them and indented like they are on the main page?

**Comment by brollb** [comment]
My bad, they are already in categories! :) Can you just update the base dockerfile with the python dependencies? I can take care of the github actions.

**Comment by dragazo** [comment]
Sure, I think since we're directly generating rst without any other tooling we just need `python3` and `python3-sphinx`. There might be an issue with the style not being installed; I can't remember if it came with the core sphinx package. So if that fails I might need to add one more dep.

**Comment by dragazo** [comment]
Most recent commit refactors everything to be more maintainable and while I was at it I switched over to using the json meta model idea you mentioned. I haven't added tests for it yet, but it structures everything we need to where it would be easy to test.

File: bin/build.js
Comment (brollb) [inline bin/build.js:121]: If you provide a default value for `str`, you can probably simplify this to just `str.trim()`.

**Review by brollb — APPROVED** [review]
This looks good. As soon as the tests are passing and the docs are generated correctly using the docker image, go ahead and merge it! (Preferably, squash and merge w/ a meaningful commit message that references the issue it closes and the PR #.)

I made one minor suggestion inline. Also, it might be nice to use ejs (supported by [lodash](https://lodash.com/docs/4.17.15#template)) rather than using your own custom regex for templating.

Nice work, @dragazo!

File: bin/build.js
Comment (dragazo) [reply bin/build.js:121]: Yeah, I tried that initially but apparently sometimes it's null

**Pull request merged**
